rfcs/45-stageless.md

```diff
@@ -0,0 +1,761 @@
+# Feature Name: `stageless`
+
+## Summary
+
+Bevy's current scheduling architecture (as of 0.7) has a large number of issues due to internal complexity.
+Certain concepts and behaviors are unnecessarily fused together and certain building blocks—stages, run criteria, states—are presented as independent but actually have significant internal coupling.
+These properties frequently lead users into encounters with unexpected side effects and indecipherable errors.
+
+It's become a tangled mess.
+
+This is a holistic redesign that decouples system storage from system scheduling, makes system labels more useful, slims down the app builder API, leverages exclusive systems to handle complex control flow (gives a clear purpose to everything in general), and more.
+
+## Motivation
+
+The fundamental challenges with the current stage-centered scheduling model are [numerous](https://github.com/bevyengine/bevy/discussions/2801).
+Of particular note:
+
+- Stages entangle ordering and command processing.
+- Plugin authors have no tools to export their systems other than wrapping them in stages, but the scheduling of imported stages can't be controlled by the user.
+- Systems and system sets cannot have multiple run criteria.
+- There's no clear pattern for users to architect turn-based games.
+- There's no clear pattern for users to make a state "span across" multiple stages with our state machine model.
+- Our state machine model (a stack) doesn't enable enough use cases to warrant its complexity.
+- There are just too many types and methods. (looking at you, "`.add_system_set_to_startup_stage`")
+
+Unfortunately, all of these problems are deeply interwoven. Despite our best efforts, fixing them incrementally at either the design or implementation stage is impossible, as it results in myopic architecture choices and terrible technical debt.
+
+## User-facing explanation
+
+### Scheduling terminology and overview
+
+Let's define some terms so that hopefully we're all on the same page.
+
+- **system**: a stateful instance of a function that accesses data stored in a world
+- **system set**: a system label that represents a group of systems and other sub-sets
+- **condition**: a special kind of system that guards the execution of a system (set)
+- **dependency**: a system (set) that must complete before the system (set) of interest can run
+- **scheduler**: the programmer, who is attempting to control when and under what conditions systems run
+- **schedule** (verb): specify when and under what conditions systems run
+- **schedule** (noun): the executable form of a system set
+- **executor**: executes a schedule on a world
+- **"ready"**: when a system is no longer waiting for dependencies to complete
+- **"complete"**: when a system has finished running or was skipped (because a condition didn't pass)
+- **"incompatible"**: when two systems have conflicting access to some data (would be data race if they ran at the same time)

```

Comment:
these terms aren't being used in the user facing explanation section and should either be moved to implementation or just removed.